### PR TITLE
Add VirtualMachine foreground finalizer

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
Currently, when deleting a vm which has an associated vmi,
the vm object is immediately removed, while the vmi object wait
for the pod deletion first.
This can cause a misunderstanding in case the vmi is not deleted
due to some errors. For example, the subsequent creation of a vm
with the same name results in the adoption of the pending vmi.
Adding a finalizer on the vm object prevents new vm creation with
the same name to be avoided, since the original vm object remains
in a terminating state, until the vmi object is gone.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://bugzilla.redhat.com/show_bug.cgi?id=2132873

**Special notes for your reviewer**:

**Release note**:

```release-note
Added foreground finalizer to  virtual machine
```

### CODE REVIEW

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vm-mutator.go
Comment (rmohr) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vm-mutator.go:74]: I think it would be best to consider the upgrade case where  the VMs already exist. That would also be in line  with the VMI controller which adds the finalizer as one of the first operation. 

This should "auto-fix" existing VMs and for new VMs we add the finalizer before we create the VMI.

What do you think?

Comment (fossedihelm) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vm-mutator.go]: Thanks for the suggestion and the offline explanation! I pushed the changes to address this! PTAL!

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go:2238]: `stopVMI` checks itself if vmi is nil, but I agree with you that `vmi == nil` should be done first. Thanks

**Comment by fabiand** [comment]
IIUIC then this will make VM deletions foreground by default. is this correct?

Thus for VMs with a long deletion timespan, this will block the deletion call for long, is this correct?

**Comment by fossedihelm** [comment]
/hold
Pushed a test commit

**Comment by fossedihelm** [comment]
Hey @rmohr! I pushed the requested changes. Can you have a look? 
I also pushed a fix for the storage failures. I want to point you on the second commit ([bfbf5cd74ea6a63239023830826ea6204a77db75](https://github.com/kubevirt/kubevirt/pull/8937/commits/bfbf5cd74ea6a63239023830826ea6204a77db75)) and a feedback from your side will be more than appreciated. Thanks

File: pkg/virt-controller/watch/vm_test.go
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:164]: Can you please use the const `virtualMachineControllerFinalize` here instead of the string literal?

File: pkg/virt-controller/watch/vm_test.go
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:4061]: WDYT about adding the finalizer only if `started == true`? I think it simulates real life better, right?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (iholder101) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:107]: This also assumes that there's only one finalizer relevant for VMs, as if we would have `anotherFinalizer` in the future, the length of `vm.Finalizers` could then be 1 therefore it'll prevent the controller from removing the controller finalizer if e.g. the data source doesn't exist.

In addition, this basically means that every VM update is legal when a VM is about to be deleted. Which validations are actually problematic here? If it's only the data source validation, then maybe we can define this as an exception especially for that case.

WDYT?

File: pkg/virt-controller/watch/vm_test.go
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:158]: Please use the const here as well

File: pkg/virt-controller/watch/vm_test.go
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:1408]: nit: you can delete this comment and expect the finalizer to be an element of vm.Finalizers. This will both serve as documentation and provide a guarantee.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (fossedihelm) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:107]: > This also assumes that there's only one finalizer relevant for VMs, as if we would have `anotherFinalizer` in the future, the length of `vm.Finalizers` could then be 1 therefore it'll prevent the controller from removing the controller finalizer if e.g. the data source doesn't exist.

This is absolutely true! :)
 
> In addition, this basically means that every VM update is legal when a VM is about to be deleted. Which validations are actually problematic here? If it's only the data source validation, then maybe we can define this as an exception especially for that case.
> 
> WDYT?

This was exactly my doubt! I definitely encountered the problem with the dataSouce, but I suppose that there should be something else (? maybe?) that could cause the same issue! In my mind I though that, since this will be the last step before the vm disappear it should be fine!
But I am not sure, for this I want your feedback :)

I can also add some information: `sig-storage` lanes was always failing, so for sure the dataSource will always create the issue. But there was also a `sig-compute` failure once (https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/8937/pull-kubevirt-e2e-k8s-1.23-sig-compute/1604848908272406528), but since there are no failing test (but only afterSuite), I am not able to find which one creates the pending VM.

File: pkg/virt-controller/watch/vm_test.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm_test.go:4061]: The finalizer will be added for both started and stopped vm, so from a real life simulation perspective IMHO is better to add it always :slightly_smiling_face: Let me know, Thanks

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (iholder101) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:107]: In general I think that seeing an error on a validation that's failing while the VMI is intended to be deleted is relatively easy to debug. A wrong and non-validated VM modification could lead to some weird side-effects that might be very hard to debug. So if we're talking only about the data source validation - I think it's better to move the `if` to there only. If many validations do not apply to a case of a VM that's being deleted and the whole validation isn't relevant that's a different story IMO.

@rmohr would appreciate your thoughts here as well

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go]: Starting from your suggestion I preferred to extract a `getPatchFinalizerOps` function that can be used in `removeVMIFinalizer` too. Thanks!

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go]: :+1:  Thanks! For the moment I extended the `add` and `remove` functions taking the finalizer you want to delete as parameter. This will help future extensions.

File: pkg/virt-controller/watch/vm_test.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm_test.go]: Done

File: pkg/virt-controller/watch/vm_test.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm_test.go]: Done

File: pkg/virt-controller/watch/vm_test.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm_test.go:1408]: Done

**Comment by fossedihelm** [comment]
Thanks @iholder101 for your detailed review! PTAL

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (fossedihelm) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: Moved the check on ds

File: pkg/virt-controller/watch/vm.go
Comment (iholder101) [reply pkg/virt-controller/watch/vm.go]: Awesome! Thanks!

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (iholder101) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:223]: nit: I know that data volumes is the only subject being tested in this function, but more validations can be added in the future. WDYT about something like:

```go
if vm.DeletionTimestamp == nil || len(vm.Finalizers) > 0 {
   for idx, dataVolume := range vm.Spec.DataVolumeTemplates {
      // ...
   }
}
```

but maybe I'm over reacting here :)

**Review by iholder101 — COMMENTED** [review]
Thanks @fossedihelm!
Left one nit :)

/lgtm

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (fossedihelm) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: Thanks for the suggestion! Done! PTAL :)

**Review by jean-edouard — COMMENTED** [review]
/approve

File: pkg/virt-controller/watch/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:1707]: I would like us to avoid unneeded `DeepCopy` at least in the hot path. Let's just iterate over the `Finalizers`.

Comment (iholder101) [reply pkg/virt-controller/watch/vm.go:1707]: Nice catch!

A bit off-topic, but now that golang has generics I would have expected that general purpose copy/equality functions would be introduced, if not in the standard library then at least by some external one. Is there anything like that out there that we can use in Kubevirt?

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go]: @xpivarc PTAL

File: pkg/virt-controller/watch/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:321]: Does this mean we have computed the status below on "old" state previously?

File: pkg/virt-controller/watch/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:1691]: This seems to be a general function but this log seems to be very specific. Maybe you want to include the finalizer we are removing?

File: pkg/virt-controller/watch/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:1708]: Same as above

File: pkg/virt-controller/watch/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:1716]: Nit: This can be inlined.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:220]: Is it enough to have the first part of the condition?

**Review by xpivarc — COMMENTED** [review]
I have a few more questions and some suggestions. I think it would be great to exercise this by functional test, wdyt?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (fossedihelm) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:220]: Even better with an early return :)
```
if vm.DeletionTimestamp != nil {
	return causes, nil
}
```
Thanks

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go:1716]: :+1:

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go:1691]: Leftover from initial implementation: Thanks!

File: pkg/virt-controller/watch/vm.go
Comment (fossedihelm) [reply pkg/virt-controller/watch/vm.go:321]: Generally no, because in the `sync` method we do not update the vm, but in some cases yes https://github.com/kubevirt/kubevirt/blob/eeb7ea2ab68a8bb44eed0f24a1ced0ba1d4a411b/pkg/virt-controller/watch/vm.go#L2295

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (awels) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:217]: I would say we only need it for create. We are validating that if we are cloning from some source, the user creating the VM has permissions to clone. Note that CDI will do the same check and reject the creation of the DVs if not allowed (Just have to make sure which user is creating the DVs, might be the kubevirt SA, not 100% sure)

**Comment by fossedihelm** [comment]
@xpivarc I pushed the changes to address the comments, can you have a look?
@jean-edouard since things are changed, do you want to have another look?
Thanks

**Review by xpivarc — APPROVED** [review]

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:2272]: This can fail. You need to use patch or `Eventually`

**Comment by xpivarc** [comment]
/hold 
Helping Tide.
The storage lane is currently broken. https://github.com/kubevirt/kubevirt/pull/9135 should fix this.
Feel free to remove the hold once this is merged.

**Comment by fossedihelm** [comment]
@xpivarc I changed the `Update` method with `Patch`. Can you have a look again? Thanks

**Review by xpivarc — APPROVED** [review]
Thanks! and I apologize for the partial reviews.

**Pull request merged**
